spring-cloud-gcp-autoconfigure/src/test/java/org/springframework/cloud/gcp/autoconfigure/pubsub/it/PubSubChannelAdaptersIT.java

```diff
@@ -0,0 +1,231 @@
+
+	@Autowired
+	private GcpPubSubProperties pubSubProperties;
+
+	@Autowired
+	@Qualifier("inputChannel")
+	private MessageChannel inputChannel;
+
+	@Autowired
+	private PubSubInboundChannelAdapter inboundChannelAdapter;
+
+	@Autowired
+	private PubSubMessageHandler outboundChannelAdapter;
+
+	@BeforeClass
+	public static void checkEmulatorIsRunning() {
+		assumeThat(System.getenv(EMULATOR_HOST_ENVVAR_NAME)).isNotNull();
+	}
+
+	@Before
+	public void setUp() {
+		this.pubSubAdmin.createTopic("desafinado");
+		this.pubSubAdmin.createSubscription("doralice", "desafinado");
+	}
+
+	@After
+	public void tearDown() {
+		this.pubSubAdmin.deleteTopic("desafinado");
+		this.pubSubAdmin.deleteSubscription("doralice");
+	}
+
+	@Test
+	public void sendAndReceiveMessage() {
+		Map<String, Object> headers = new HashMap<>();
+		// Only String values for now..
+		headers.put("storm", "lift your skinny fists");
+		headers.put("static", "lift your skinny fists");
+		headers.put("sleep", "lift your skinny fists");
+
+		this.inputChannel.send(
+				MessageBuilder.createMessage("I am a message.",  new MessageHeaders(headers)));
+
+		Message<?> message = this.channel.receive(5000);
+		assertThat(message).isNotNull();
+		assertThat(message.getPayload()).isInstanceOf(String.class);
+		String payload = (String) message.getPayload();
+		assertThat(payload).isEqualTo("I am a message.");
+
+		assertThat(message.getHeaders().size()).isEqualTo(6);
+		assertThat(message.getHeaders().get("storm")).isEqualTo("lift your skinny fists");
+		assertThat(message.getHeaders().get("static")).isEqualTo("lift your skinny fists");
+		assertThat(message.getHeaders().get("sleep")).isEqualTo("lift your skinny fists");
+	}
+
+	@Test
+	@DirtiesContext
+	public void sendAndReceiveMessageInBytes() {
+		this.inboundChannelAdapter.setMessageConverter(null);
+		this.inputChannel.send(MessageBuilder.withPayload("I am a message.").build());
+
+		Message<?> message = this.channel.receive(5000);
+		assertThat(message).isNotNull();
+		assertThat(message.getPayload()).isInstanceOf(byte[].class);
+		String stringPayload = new String((byte[]) message.getPayload());
+		assertThat(stringPayload).isEqualTo("I am a message.");
+	}
+
+	@Test
+	@DirtiesContext
+	public void sendAndReceiveMessageManualAck() {
+		this.inboundChannelAdapter.setAckMode(AckMode.MANUAL);
+		this.inputChannel.send(MessageBuilder.withPayload("I am a message.").build());
+
+		Message<?> message = this.channel.receive(5000);
+		assertThat(message).isNotNull();
+		AckReplyConsumer acker =
+				(AckReplyConsumer) message.getHeaders().get(GcpHeaders.ACKNOWLEDGEMENT);
+		assertThat(acker).isNotNull();
+		acker.nack();
+		message = this.channel.receive(5000);
+		assertThat(message).isNotNull();
+		acker = (AckReplyConsumer) message.getHeaders().get(GcpHeaders.ACKNOWLEDGEMENT);
+		assertThat(acker).isNotNull();
+		acker.ack();
+		message = this.channel.receive(5000);
+		assertThat(message).isNull();
+	}
+
+	@Test
+	@DirtiesContext
```

Comment:
Can we consider to use `start()/stop()` on the `this.inboundChannelAdapter` instead?

I fully don't like when tests spend too much time. There are many other interesting things rather than wait for tests to pass, when we can just make them faster.

Reply:
Not sure if I understand the comment.

`start()/stop()` don't restore the default values of the fields in the object. For each test, we want a fresh instance of a bean.

Reply:
Right, they don't, but you can change adapters settings in between stop and start.
That's just doesn't look right to have `@DirtiesContext` on each test method.
Again: performance concern.

Reply:
It's only on each method that modifies the bean being integration tested (i.e., `PubSubInboundChannelAdapter`).

If I understand your suggestion correctly, it's that `start()/stop()` do restore the bean to its original default config? I wonder if that's a legitimate expectation? If a user configures a bean and then stops and starts, they're probably expecting the bean to have the same configuration?

Restoring the configuration on start()/stop() just because of tests doesn't sound right, let me know if I'm missing anything.

Reply:
The problem with the `@DirtiesContext` on each method that it fully restart the whole application context. I found it as very costly and would be better to just change properties of the bean in between stop and start. Right, you have to restore properties manually before starting the bean again. 

That's just doesn't feel reasonable to waste time just because you are lazy to restore bean state manually for each test-case.

Just IMO again. Although you should count that CI server is on our side, so respect its resources is friendly way to go as well :smile:

Reply:
Aah, you mean the test's `@Before (start())/@After (stop())`?

That could be a better place to reset the properties. But it would require that we duplicate code that is currently on `PubSubInboundChannelAdapter`'s constructor, because we don't call the constructor, the Spring context does.

I could add something like

```
		StringMessageConverter stringMessageConverter = new StringMessageConverter();
		stringMessageConverter.setSerializedPayloadClass(String.class);
		this.inboundChannelAdapter.setMessageConverter(stringMessageConverter);
```

to the test's `@Before`, but that's an ugly duplication of code...

Reply:
Another way to go may be reached with only single test method in this class.